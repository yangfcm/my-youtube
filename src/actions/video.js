import { defAxios as axios } from "../settings";
import { FETCH_VIDEOS, FETCH_VIDEO, CATCH_ERROR } from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";

/**
 * Get a set of videos based on filter,
 * e.g. videos belonging to a channel or videos belonging to a play list etc.
 */
export const fetchVideos = (filter, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          ...filter,
          part: "snippet,statistics",
          maxResults: 15,
          pageToken,
        },
      });
      dispatch({
        type: FETCH_VIDEOS,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: {
          ...e.response.data.error,
          displayMessage: DEFAULT_ERROR_MSG.FAILED_TO_FETCH_VIDEO,
        },
      });
    }
  };
};

/**
 * Get a video by id
 */
export const fetchVideo = (videoId) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          part: "snippet,statistics",
          id: videoId,
        },
      });
      dispatch({
        type: FETCH_VIDEO,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: {
          ...e.response.data.error,
          displayMessage: DEFAULT_ERROR_MSG.FAILED_TO_FETCH_VIDEO,
        },
      });
    }
  };
};
