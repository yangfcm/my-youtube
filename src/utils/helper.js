/**
 * Separate every 3 number digits by comma.
 * e.g. 12345678 => 12,345,678
 * @param {number} number
 */
export const separateNumber = number => {
  let strNum = String(number);
  let arrNum = strNum.split("");
  let arr = [];
  let part = "";
  for (let i = 0; i < arrNum.length; i++) {
    let num = arrNum[arrNum.length - i - 1];
    part = num + part;
    if ((i + 1) % 3 === 0) {
      arr.unshift(part);
      part = "";
    }
  }
  if (part.length > 0) {
    arr.unshift(part);
  }
  return arr.join(",");
};
