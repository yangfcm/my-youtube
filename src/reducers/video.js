import { FETCH_VIDEOS, FETCH_VIDEO } from "../actions/types";

export const videoReducer = (state = {}, action) => {
  switch (action.type) {
    case FETCH_VIDEOS:
      return { ...state, videos: action.payload };
    case FETCH_VIDEO:
      return { ...state, video: action.payload };
    default:
      return state;
  }
};
