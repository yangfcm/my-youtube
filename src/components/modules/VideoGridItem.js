import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import { separateNumber } from "../../utils/helper";

const VideoGridItem = ({ video }) => {
  return (
    <div>
      <div className="mb-2">
        <Link to={`/video/${video.id.videoId || video.id}`}>
          <img
            style={{ width: "100%", height: "auto" }}
            src={video.snippet.thumbnails.medium.url}
            alt={video.snippet.title}
            title={video.snippet.title}
          />
        </Link>
      </div>
      <div>
        <div
          style={{
            lineHeight: "23px",
            maxHeight: "46px",
            overflow: "hidden",
            textOverflow: "ellipsis",
          }}
          className="mb-1"
        >
          <Link
            to={`/video/${video.id.videoId || video.id}`}
            title={video.snippet.title}
          >
            {video.snippet.title}
          </Link>
        </div>
        <div className="mb-2">
          <Link to={`/channel/${video.snippet.channelId}`}>
            <span className="text-secondary">{video.snippet.channelTitle}</span>
          </Link>
        </div>
        {
          <div className="text-muted">
            {video.statistics && (
              <span>{separateNumber(video.statistics.viewCount)} Views |</span>
            )}{" "}
            {moment(video.snippet.publishedAt).fromNow()}
          </div>
        }
      </div>
    </div>
  );
};

export default VideoGridItem;
