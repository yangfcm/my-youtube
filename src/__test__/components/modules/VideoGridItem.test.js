import React from "react";
import { shallow } from "enzyme";
import { Link } from "react-router-dom";
import moment from "moment";
import VideoGridItem from "components/modules/VideoGridItem";
import { separateNumber } from "utils/helper";
import { videosResponse } from "__test__/fixtures/video";

describe("Test VideoGridItem component", () => {
  let wrapper;
  const video = videosResponse.items[0];

  beforeEach(() => {
    wrapper = shallow(<VideoGridItem video={video} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct video image", () => {
    expect(wrapper.find("img").prop("src")).toBe(
      video.snippet.thumbnails.medium.url
    );
  });

  it("should render correct links to that navigate to video and channel", () => {
    const links = wrapper.find(Link);
    expect(links.at(0).prop("to")).toBe(`/video/${video.id}`);
    expect(links.at(1).text()).toContain(video.snippet.title);
    expect(links.at(1).prop("to")).toBe(`/video/${video.id}`);
    expect(links.at(2).text()).toContain(video.snippet.channelTitle);
    expect(links.at(2).prop("to")).toBe(`/channel/${video.snippet.channelId}`);
  });

  it("should render correct view count and publish date time", () => {
    const muteText = wrapper.find("div.text-muted").text();
    expect(muteText).toContain(separateNumber(video.statistics.viewCount));
    expect(muteText).toContain(moment(video.snippet.publishedAt).fromNow());
  });
});
