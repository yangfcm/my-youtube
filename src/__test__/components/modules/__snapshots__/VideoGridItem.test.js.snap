// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test VideoGridItem component should render component 1`] = `
<div>
  <div
    className="mb-2"
  >
    <Link
      to="/video/oxoWhyS9buA"
    >
      <img
        alt="BTS (방탄소년단) 'Dynamite' Official Teaser"
        src="https://i.ytimg.com/vi/oxoWhyS9buA/mqdefault.jpg"
        style={
          Object {
            "height": "auto",
            "width": "100%",
          }
        }
        title="BTS (방탄소년단) 'Dynamite' Official Teaser"
      />
    </Link>
  </div>
  <div>
    <div
      className="mb-1"
      style={
        Object {
          "lineHeight": "23px",
          "maxHeight": "46px",
          "overflow": "hidden",
          "textOverflow": "ellipsis",
        }
      }
    >
      <Link
        title="BTS (방탄소년단) 'Dynamite' Official Teaser"
        to="/video/oxoWhyS9buA"
      >
        BTS (방탄소년단) 'Dynamite' Official Teaser
      </Link>
    </div>
    <div
      className="mb-2"
    >
      <Link
        to="/channel/UC3IZKseVpdzPSBaWxBxundA"
      >
        <span
          className="text-secondary"
        >
          Big Hit Labels
        </span>
      </Link>
    </div>
    <div
      className="text-muted"
    >
      <span>
        22,928,171
         Views |
      </span>
       
      3 months ago
    </div>
  </div>
</div>
`;
